### DISCUSSION TITLE
Linter that disallows Generics

### DISCUSSION THREAD

**Discussion by anders-swanson** [golangci/golangci-lint#3281] [Q&A]
Some users of go 1.18+ may _not_ want to use Generics in their code. Is it reasonable to add a linter that rejects code using Generics?
The use case for this is: 
Project A has multiple release branches, which have different supported go versions. Trunk uses go 1.19, but the last major release uses go 1.17. In order to ensure new features may be backported from Trunk to the last major release, the developers agree not to use new language features (like Generics) for the time being.

**Comment by ldez** [comment]
Hello,

A linter seems not the right approach, I think it's better to change your CI to use go1.17.
The compiler will do the job.

**Reply by anders-swanson** [reply]
That's a good idea - I do still want release artifacts to be built with go 1.19, but it should compile with 1.17.

**Reply by anders-swanson** [reply]
although, compilation using 1.17 may not work because of dependencies - that could be a problem.

**Reply by ldez** [reply]
if you have `go 1.17` in your `go.mod` the compiler (even with go1.19) will throw an error:

```console
$ go version
go version go1.19.2 linux/amd64
$ go run .
# github.com/golangci/sandbox
./main.go:6:8: implicit function instantiation requires go1.18 or later (-lang was set to go1.17; check go.mod)
./main.go:9:13: type parameter requires go1.18 or later (-lang was set to go1.17; check go.mod)
./main.go:9:15: predeclared any requires go1.18 or later (-lang was set to go1.17; check go.mod)
```

Note: it is not recommended to use go1.17 for production, as it is no longer maintained, I recommend upgrading to go1.19 (or at least go1.18)
